File: processing/syntactic/expressions/definitions/macro_def.py
from __future__ import annotations

from typing import List

from processing.lexical.tokens.token_type import TokenType
from processing.syntactic.expressions.atomic.name_expr import NameExpr
from processing.syntactic.expressions.block_expr import BlockExpr, BlockType
from processing.syntactic.expressions.expr import Expr, child_property
from processing.syntactic.expressions.expression_groups import DefinitionExpression
from processing.syntactic.expressions.macro_patterns import CascadePattern
from processing.syntactic.expressions.macro_patterns import MacroPattern


class MacroDef(DefinitionExpression):
    """macro_def:
       'macro' name block;
    """

    @child_property
    def name(self) -> NameExpr:
        pass

    @child_property
    def block(self) -> BlockExpr:
        pass

    @child_property
    def modifiers(self) -> List[Expr]:
        pass

    def __init__(
            self,
            parent: Expr = None,
            name: NameExpr = None,
            block: BlockExpr = None,
            modifiers: List[Expr] = None,
            patterns: List[MacroPattern] = None
    ):
        super().__init__(parent)
        self.name = name
        self.block = block
        self.modifiers = modifiers
        self.syntax = CascadePattern(*patterns)

    def parse(self) -> MacroDef:
        self.stream.eat(TokenType.keyword_macro)
        self.name = NameExpr(self).parse(must_be_simple = True)
        self.block = BlockExpr(self).parse(BlockType.default)
        return self

File: Axion.Core/Processing/Syntactic/Expressions/Definitions/MacroDef.cs
using System.Collections.Generic;
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Atomic;
using Axion.Core.Processing.Syntactic.Expressions.Patterns;
using Magnolia.Attributes;
using static Axion.Specification.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions.Definitions;

/// <summary>
///     <code>
///         macro-def:
///             'macro' simple-name syntax-description scope;
///     </code>
/// </summary>
[Branch]
public partial class MacroDef : Node, IDefinitionExpr {
    [Leaf] Token? kwMacro;
    [Leaf] NameExpr? name;
    [Leaf] ScopeExpr scope = null!;
    [Leaf] CascadePattern syntax = null!;

    public Dictionary<string, string> NamedSyntaxParts { get; } = new();

    internal MacroDef(Node parent) : base(parent) { }

    public MacroDef Parse() {
        // TODO: find code, that can be replaced with macro by patterns
        // Example:
        // ========
        // macro post-condition-loop (
        //   'do',
        //   scope: Scope,
        //   ('while' | 'until'),
        //   condition: Infix
        // )
        //   if syntax[2] == 'while'
        //     condition = {{ not $condition }}
        // 
        //   return {{
        //     while true {
        //       $scope
        //       if $condition {
        //         break
        //       }
        //     }
        //   }}
        KwMacro = Stream.Eat(KeywordMacro);
        Name    = new NameExpr(this).Parse(true);
        Syntax  = new CascadePattern(this);
        // EBNF-based syntax definition
        if (Stream.MaybeEat(OpenParenthesis)) {
            Syntax.Parse();
            Stream.Eat(CloseParenthesis);
        }

        Scope = new ScopeExpr(this).Parse();
        return this;
    }
}
